librz/core/cmd/cmd_print.c

```diff
@@ -6599,3 +6382,219 @@ RZ_IPI RzCmdStatus rz_print_byte_bitstream_handler(RzCore *core, int argc, const
 	free(str_buf);
 	return RZ_CMD_STATUS_OK;
 }
+
+typedef struct {
+	ut64 size;
+	ut64 repeat;
+	bool useBytes;
+} PrintValueOptions;
+
+static void print_value_single(RzCore *core, PrintValueOptions *opts, ut64 address, ut64 value, RzCmdStateOutput *state) {
+	switch (state->mode) {
+	case RZ_OUTPUT_MODE_RIZIN:
+		rz_cons_printf("f pval.0x%08" PFMT64x " @ %" PFMT64d "\n", address, value);
+		break;
+	case RZ_OUTPUT_MODE_STANDARD:
+		switch (opts->size) {
+		case 1:
+			rz_cons_printf("0x%02" PFMT64x "\n", value);
+			break;
+		case 2:
+			rz_cons_printf("0x%04" PFMT64x "\n", value);
+			break;
+		case 4:
+			rz_cons_printf("0x%08" PFMT64x "\n", value);
+			break;
+		case 8:
+			rz_cons_printf("0x%016" PFMT64x "\n", value);
+			break;
+		default:
+			rz_warn_if_reached();
+			break;
+		}
+		break;
+	case RZ_OUTPUT_MODE_JSON: {
+		// TODO: Use API instead of the command
+		char *str = rz_core_cmd_str(core, "ps");
+		rz_str_trim(str);
+		char *p = str;
+		if (p) {
+			while (*p) {
+				if (*p == '\\' && p[1] == 'x') {
+					memmove(p, p + 4, strlen(p + 4) + 1);
+				}
+				p++;
+			}
+		}
+		pj_o(state->d.pj);
+		pj_k(state->d.pj, "value");
+		switch (opts->size) {
+		case 1:
+		case 2:
+			pj_i(state->d.pj, value);
+			break;
+		case 4:
+		case 8:
+			pj_n(state->d.pj, value);
+			break;
+		default:
+			rz_warn_if_reached();
+			break;
+		}
+		pj_ks(state->d.pj, "string", str);
+		pj_end(state->d.pj);
+		break;
+	}
+	default:
+		rz_warn_if_reached();
+		break;
+	}
+}
+
+static bool print_value(RzCore *core, PrintValueOptions *opts, RzCmdStateOutput *state) {
+	ut64 old_at = core->offset;
+	ut8 *block = core->block;
+	int blocksize = core->blocksize;
+	ut8 *block_end = core->block + blocksize;
+
+	if (block + 8 >= block_end) {
+		RZ_LOG_ERROR("core: block is truncated.\n");
+		return false;
+	}
```

Comment:
I think the block concept can be removed and we can use `rz_io_nread_at` instead. In any case, we can probably avoid the `rz_core_seek` in the do/while loop below and just keep track of `at` manually (you can use `ps @ {addr}` in the print_value_single function).

Reply:
I would prefer to postpone this until converting to the API calls, after `ps` are converted to the rzshell and a better API.